mu/modes/snek.py

```diff
@@ -0,0 +1,394 @@
+        "SERVO4",
+        "SIGNAL1",
+        "SIGNAL2",
+        "SIGNAL3",
+        "SIGNAL4",
+        "SIGNAL5",
+        "SIGNAL6",
+        "SIGNAL7",
+        "SIGNAL8",
+        "SWITCH",
+        "TEMP",
+        "TX",
+        "abs_tol",
+        "curses",
+        "eeprom",
+        "exit",
+        "math",
+        "neopixel",
+        "off",
+        "on",
+        "onfor",
+        "pulldown",
+        "pullnone",
+        "pullup",
+        "random",
+        "read",
+        "rel_tol",
+        "reset",
+        "round",
+        "setleft",
+        "setpower",
+        "setright",
+        "stdscr",
+        "stopall",
+        "sys",
+        "talkto",
+        "temperature",
+        "time",
+    )
+
+    def stop(self):
+        self.remove_repl()
+
+    def actions(self):
+        """
+        Return an ordered list of actions provided by this module. An action
+        is a name (also used to identify the icon) , description, and handler.
+        """
+        buttons = [
+            {
+                "name": "serial",
+                "display_name": _("Serial"),
+                "description": _("Open a serial connection to your device."),
+                "handler": self.toggle_repl,
+                "shortcut": "CTRL+Shift+U",
+            },
+            {
+                "name": "flash",
+                "display_name": _("Put"),
+                "description": _("Put the current program to the device."),
+                "handler": self.put,
+                "shortcut": "CTRL+Shift+P",
+            },
+            {
+                "name": "getflash",
+                "display_name": _("Get"),
+                "description": _("Get the current program from the device."),
+                "handler": self.get,
+                "shortcut": "CTRL+Shift+G",
+            },
+        ]
+        if CHARTS:
+            buttons.append(
+                {
+                    "name": "plotter",
+                    "display_name": _("Plotter"),
+                    "description": _("Plot incoming REPL data."),
+                    "handler": self.toggle_plotter,
+                    "shortcut": "CTRL+Shift+P",
+                }
+            )
+        return buttons
+
+    def put(self):
+        """
+        Put the current program into the device memory.
+        """
+        logger.info("Downloading code to target device.")
+        # Grab the Python script.
+        tab = self.view.current_tab
+        if tab is None:
+            # There is no active text editor.
+            message = _("Cannot run anything without any active editor tabs.")
+            information = _(
+                "Running transfers the content of the current tab"
+                " onto the device. It seems like you don't have "
+                " any tabs open."
+            )
+            self.view.show_message(message, information)
+            return
+        python_script = tab.text()
+        if python_script[-1] != "\n":
```

Comment:
Raises IndexError if the document is empty, should be something like `if not python_script or python_script[-1] != "\n"`.

Reply:
thanks. I added a 'not python_script or' to the beginning of the condition.